src/sidebar/store/test/use-store-test.js

```diff
@@ -16,90 +18,262 @@ const reducer = (state = initialState, action) => {
+      assert.called(warnOnce);
+      assert.match(warnOnce.firstCall.args[0], /changes every time/);
+    });
+
+    it('unsubscribes when the component is unmounted', () => {
+      const unsubscribe = sinon.stub();
+      testStore.subscribe = sinon.stub().returns(unsubscribe);
+
+      const wrapper = mount(<TestComponent />);
+
+      assert.calledOnce(testStore.subscribe);
+      wrapper.unmount();
+      assert.calledOnce(unsubscribe);
     });
-    const BuggyComponent = () => {
-      // The result of the callback is an object with an `aValue` property
-      // which is a new array every time. This causes unnecessary re-renders.
-      useStore(() => ({ aValue: [] }));
-      return null;
-    };
-    mount(<BuggyComponent />);
-    assert.called(warnOnce);
-    assert.match(warnOnce.firstCall.args[0], /changes every time/);
   });
 
-  it('unsubscribes when the component is unmounted', () => {
-    const unsubscribe = sinon.stub();
-    testStore.subscribe = sinon.stub().returns(unsubscribe);
+  describe('useStoreProxy', () => {
+    let store;
+    let renderCount;
+
+    beforeEach(() => {
+      renderCount = 0;
+      store = createStore([thingsModule]);
+
+      store.addThing('foo');
+      store.addThing('bar');
+
+      $imports.$mock({
+        '../util/service-context': {
+          useService: name => (name === 'store' ? store : null),
+        },
+      });
+    });
+
+    function renderTestComponent() {
+      let proxy;
+
+      const TestComponent = () => {
+        ++renderCount;
+        proxy = useStoreProxy();
+
+        return <div>{proxy.thingCount()}</div>;
+      };
+
+      const wrapper = mount(<TestComponent />);
+      return { proxy, wrapper };
+    }
+
+    it('returns a proxy for the store', () => {
+      const addThingSpy = sinon.spy(store, 'addThing');
 
-    const wrapper = mount(<TestComponent />);
+      const { proxy } = renderTestComponent();
 
-    assert.calledOnce(testStore.subscribe);
-    wrapper.unmount();
-    assert.calledOnce(unsubscribe);
+      assert.ok(proxy);
+
+      // Test proxied selector method.
+      assert.deepEqual(proxy.getThing('foo'), { id: 'foo' });
+      assert.deepEqual(proxy.getThing('bar'), { id: 'bar' });
+
+      // Test proxied action dispatch.
+      proxy.addThing('baz');
+      assert.calledWith(addThingSpy, 'baz');
+    });
+
+    it('proxies non-function properties of the store', () => {
+      store.someString = 'foobar';
+
+      const { proxy } = renderTestComponent();
+
+      assert.equal(proxy.someString, 'foobar');
+    });
+
+    it('records and caches selector method calls', () => {
+      const getThingSpy = sinon.spy(store, 'getThing');
+
+      const { proxy } = renderTestComponent();
+
+      proxy.getThing('foo');
+      proxy.getThing('foo');
+
+      assert.calledWith(getThingSpy, 'foo');
+      assert.calledOnce(getThingSpy);
+      getThingSpy.resetHistory();
+
+      proxy.getThing('bar');
+      proxy.getThing('bar');
```

Comment:
Do you feel like a test like this sufficiently exercises the path of "the component re-renders and re-invokes its various proxied store selector methods and we can see that those selectors are not re-invoked because their cache entries are still valid"?

Reply:
It does cover that functionality, but only in a minimal way. I think that's adequate for now but we might want to extend it in future.